From dcf398ebfb3010768b96bdf53911e513db142ecf Mon Sep 17 00:00:00 2001
From: Grzegorz Jaszczyk <jaz@semihalf.com>
Date: Mon, 2 Mar 2020 13:05:18 +0100
Subject: [PATCH 1122/1239] sar: mvebu: use smc calls for dfx register-set
 access

This change will be needed after dfx register set will be secured by the
trusted firmware.

Change-Id: I4cbd685976d2faf0ca98872763af85e2c4db0143
Signed-off-by: Grzegorz Jaszczyk <jaz@semihalf.com>
Reviewed-on: https://sj1git1.cavium.com/25052
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Kostya Porotchkin <kostap@marvell.com>
---
 drivers/misc/mvebu_sar/ap807_sar.c | 24 +++++++++++++++++++-----
 1 file changed, 19 insertions(+), 5 deletions(-)

diff --git a/drivers/misc/mvebu_sar/ap807_sar.c b/drivers/misc/mvebu_sar/ap807_sar.c
index eb306cbf01..5ea26e60e2 100644
--- a/drivers/misc/mvebu_sar/ap807_sar.c
+++ b/drivers/misc/mvebu_sar/ap807_sar.c
@@ -129,11 +129,17 @@ static const u32 pll_freq_tbl[16][4] = {
 
 static u32 sar_get_clock_freq_mode(struct udevice *dev)
 {
-	u32 i;
+	u32 i, clock_freq;
 	struct dm_sar_pdata *priv = dev_get_priv(dev);
+	int ret;
+
+	ret = mvebu_dfx_sread(&clock_freq, (uintptr_t)priv->sar_base);
+	if (ret != SMCCC_RET_SUCCESS) /* try legacy read */
+		clock_freq = (readl(priv->sar_base));
+
+	clock_freq = (clock_freq & SAR_CLOCK_FREQ_MODE_MASK) >>
+		      SAR_CLOCK_FREQ_MODE_OFFSET;
 
-	u32 clock_freq = (readl(priv->sar_base) & SAR_CLOCK_FREQ_MODE_MASK) >>
-			  SAR_CLOCK_FREQ_MODE_OFFSET;
 	for (i = 0; i < 7; i++) {
 		if (pll_freq_tbl[i][3] == clock_freq)
 			return i;
@@ -172,8 +178,12 @@ static int ap807_sar_dump(struct udevice *dev)
 	u32 reg, val;
 	struct sar_info *sar;
 	struct dm_sar_pdata *priv = dev_get_priv(dev);
+	int ret;
+
+	ret = mvebu_dfx_sread(&reg, (uintptr_t)priv->sar_base);
+	if (ret != SMCCC_RET_SUCCESS) /* try legacy */
+		reg = readl(priv->sar_base);
 
-	reg = readl(priv->sar_base);
 	printf("AP807 SAR register 0 [0x%08x]:\n", reg);
 	printf("----------------------------------\n");
 	sar = ap807_sar_0;
@@ -183,7 +193,11 @@ static int ap807_sar_dump(struct udevice *dev)
 		sar++;
 	}
 
-	reg = readl(priv->sar_base + AP807_SAR_1_REG);
+	ret = mvebu_dfx_sread(&reg,
+			      (uintptr_t)(priv->sar_base + AP807_SAR_1_REG));
+	if (ret != SMCCC_RET_SUCCESS) /* try legacy */
+		reg = readl(priv->sar_base + AP807_SAR_1_REG);
+
 	printf("\nAP807 SAR register 1 [0x%08x]:\n", reg);
 	printf("----------------------------------\n");
 	sar = ap807_sar_1;
-- 
2.29.0

